File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.java
class Solution {
    public int yespalin(String s, int start, int end, int[][] dp) {
        if (start >= end)
            return 1;
        if (dp[start][end] != -1)
            return dp[start][end];
        if (s.charAt(start) != s.charAt(end)) {
            dp[start][end] = 0;
            return 0;
        }
        return dp[start][end] = yespalin(s, start + 1, end - 1, dp);
    }

    public String longestPalindrome(String s) {
        int n = s.length();
        int maxlen = 0;
        StringBuilder reqstr = new StringBuilder();
        
        int[][] dp = new int[n][n];
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                dp[i][j] = -1;
            }
        }

       
        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {
                
                if (yespalin(s, i, j, dp) == 1) {
                    if (j - i + 1 > maxlen) {
                        reqstr = new StringBuilder(s.substring(i, j + 1)); 
                        maxlen = Math.max(maxlen, j - i + 1);
                    }
                }
            }
        }
        return reqstr.toString();
    }
}

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.length();
        if (n <= 1) return s;

        int maxlen = 1;
        int start = 0;

        for (int i = 0; i < n; ++i) {
            int l = i;
            int r = i;
            while (l >= 0 && r < n && s[l] == s[r]) {
                if (r - l + 1 > maxlen) {
                    maxlen = r - l + 1;
                    start = l;
                }
                l--;
                r++;
            }
        }

        for (int i = 0; i < n - 1; ++i) {
            int l = i;
            int r = i + 1;
            while (l >= 0 && r < n && s[l] == s[r]) {
                if (r - l + 1 > maxlen) {
                    maxlen = r - l + 1;
                    start = l;
                }
                l--;
                r++;
            }
        }

        return s.substr(start, maxlen);
    }
};